File: frontend/src/types/pageStreak.ts
// frontend/src/types/pageStreak.ts
export type RunStatus = 'not_started' | 'in_progress' | 'completed';

export interface RosterEntry {
  killer: string;
  status: RunStatus;
  attempt: number;
  current_page: number;
  best_page: number;
  page_count: number;
  avatar_local_path?: string | null;
}

export interface HistoryEntry {
  attempt: number;
  page_number: number;
  perks: string[];
  result: 'win' | 'loss';
  timestamp: string;
  triggered_by: 'player' | 'inactivity';
}

export interface PageStreakRun {
  id: number;
  killer: string;
  status: 'in_progress' | 'completed';
  attempt: number;
  current_page: number;
  best_page: number;
  pages: string[][];
  page_count: number;
  snapshot_at: string;
  history: HistoryEntry[];
  perk_icons: Record<string, string>;
  killer_avatar?: string | null;
}

export interface PoolSummary {
  pool_size: number;
  page_count: number;
  perks_per_page: number;
}

export interface PageStreakMatchLog {
  id: number;
  run_id: number;
  killer: string;
  attempt: number;
  page_number: number;
  perks: string[];
  result: 'win' | 'loss';
  timestamp: string;
  triggered_by: 'player' | 'inactivity';
}

export interface PageStreakStats {
  total_matches: number;
  wins: number;
  losses: number;
  win_rate: number;
  recent_logs: PageStreakMatchLog[];
}

File: backend/app/schemas/page_streak.py
# backend/app/schemas/page_streak.py
from typing import TypedDict


class PageStreakPageLogDict(TypedDict):
    id: int
    run_id: int
    attempt: int
    page_number: int
    perks: list[str]
    result: str
    timestamp: str | None
    triggered_by: str


class PageStreakStatsLog(PageStreakPageLogDict):
    killer: str | None


class PageStreakHistoryEntry(TypedDict):
    attempt: int
    page_number: int
    perks: list[str]
    result: str
    timestamp: str | None
    triggered_by: str


class PageStreakRunDict(TypedDict):
    id: int
    killer: str
    status: str
    attempt: int
    current_page: int
    best_page: int
    pages: list[list[str]]
    page_count: int
    pool_frozen: bool
    snapshot_at: str | None
    history: list[PageStreakHistoryEntry]


class PageStreakRunState(PageStreakRunDict):
    """What every page streak run endpoint returns: the run plus its artwork."""
    perk_icons: dict[str, str]
    killer_avatar: str | None
